### ISSUE TITLE
Custom tangent types for gradient computations

### ISSUE THREAD

**Issue by jheek** [google/jax#918]
I would like to encapsulate a Model which has state and paramaters into a an object

```
class Model():
  params: P
  state: S
```

Currently I can use `jax.tree_util.register_pytree_node` to register this type as something that can be lifted into a trace context. However, for something like `grad(loss)(model)` I would like to receive only the gradients w.r.t. the parameters so I would like the tangent type of `Model` to be `P`.

Perhaps we could have something like `jax.register_tangent(Model, \lambda m: m.params)` although I would prefer something a bit more pythonic like a `tangent` method on `Model`.

**Comment by nkitaev** [comment]
#844 is potentially related, in the sense that auxiliary data for backprop can be implemented by defining a custom *cotangent* type.

**Comment by jheek** [comment]
Perhaps this could also be implemented using a [lens](http://hackage.haskell.org/package/lens) utility?   Currently I manually define a lens to extract the variables I want to take a gradient of and put the traced variables back into the object using that same lens.
